### PR TITLE
Swagger

### PR DESCRIPTION
**Pull request by mattgrill**
Waterwheel now parses Swagger (OpenAPI) documentation. The structure of how resources are stored has changed, _see the documentation_. The loss of `. addResources()` was prompted by the complexity of manually creating a custom resource object to add.

### CODE REVIEW

**Comment by prestonso** [comment]
Performing a code review!

total-recall-1

File: lib/entity.js [hunk 1]
```diff
@@ -30,86 +30,94 @@ module.exports = class Entity extends Request {
...
-  getFieldData(format = 'json'){
-    if (!this.options.methods.OPTIONS) {
-      return Promise.reject(`No available OPTIONS path for ${this.options.bundle}.`);
-    }
-    return this.issueRequest(methods.get, `${this.options.methods.OPTIONS.replace(this.options.methods.OPTIONS.match(/\{.*?\}/), this.options.bundle)}?_format=${format}`)
-      .then(bundleMetadata => {
-        this.metadata = bundleMetadata;
-        return Promise.resolve(bundleMetadata);
-      });
+  checkRequiredFields(body) {
```
Comment (prestonso) [inline lib/entity.js:126]: Nit: no docblock

**Comment by prestonso** [comment]
Reviewed some, but need to resume later.

**Comment by infiniteluke** [comment]
This addresses #49 and should be accepted in favor of #50. Thanks @mattgrill 🥇

File: lib/entity.js [hunk 2]
```diff
@@ -123,51 +139,45 @@ module.exports = class Entity extends Request {
...
-              return fieldKey;
-            }
-            return false;
-          }).filter(Boolean);
-
-        if (nonMatchedFields.length){
-          return reject(new Error(`The ${nonMatchedFields.length > 1 ? 'fields' : 'field'}, ${nonMatchedFields.join(', ')}, ${nonMatchedFields.length > 1 ? 'are' : 'is'} not included within the bundle, ${this.options.bundle}.`));
+  setFields(identifier, fields) {
+    const availableFieldKeys = Object.keys(this.options.metadata.properties);
+    // const nonMatchedFields = (Array.isArray(fields) ? fields : [fields])
```
Comment (infiniteluke) [inline lib/entity.js]: Remove comments?

File: lib/entity.js [hunk 3]
```diff
@@ -123,51 +139,45 @@ module.exports = class Entity extends Request {
...
+
+      postBody[fieldNameToSet].push(fieldValueWrapper);
     });
-  }
 
+    postBody.type = [{
+      target_id: this.options.bundle,
+      target_type: `${this.options.entity}_type`
+    }];
+    // return Promise.resolve(postBody);
```
Comment (infiniteluke) [inline lib/entity.js]: Commented code

File: lib/jsonapi.js [hunk 4]
```diff
@@ -22,7 +23,7 @@ module.exports = class JSONAPI extends Request {
  */
   get(resource, params, id = false) {
     const format = 'api_json';
-    const url = `/api/${resource}${id ? `/${id}` : ''}?_format=${format}${Object.keys(params).length ? `&${qs.stringify(params, {indices: false})}` : ''}`;
+    const url = `/${this.jsonapiPrefix}/${resource}${id ? `/${id}` : ''}?_format=${format}${Object.keys(params).length ? `&${qs.stringify(params, {indices: false})}` : ''}`;
```
Comment (infiniteluke) [inline lib/jsonapi.js:26]: yey

File: lib/jsonapi.js [hunk 5]
```diff
@@ -38,7 +39,38 @@ module.exports = class JSONAPI extends Request {
...
+   * PATCH jsonapi
+   *
+   * @param {string} resource
+   *   The relative path to fetch from the API.
+   * @param  {object} body
+   *   JSON data sent to Drupal
+   * @return {promise}
+   *   Resolves when the request is fulfilled, rejects if there's an error.
+   */
+  patch(resource, body) {
```
Comment (infiniteluke) [inline lib/jsonapi.js:63]: yeeeeeeey

File: lib/swagger.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+            parameters: this.swagger.paths[path][method].parameters
+          };
+
+          // Setup bundle properties.
+          this.entities[bundle].properties = this.entities[bundle].properties ?
+            this.entities[bundle].properties :
+            bundleData.properties;
+
+          // Setup required fields.
+          this.entities[bundle].requiredFields = this.entities[bundle].requiredFields ?
```
Comment (infiniteluke) [inline lib/swagger.js:45]: This is awesome

**Review by infiniteluke — APPROVED** [review]
The DX gains from validating swagger methods/required fields is huge.

**Review by infiniteluke — APPROVED** [review]
lgtm!

**Pull request merged**